proxy/src/transport/prefixed.rs

```diff
@@ -0,0 +1,86 @@
+#![allow(dead_code)] // TODO: Actually use this.
+
+use std::{cmp, fmt::Debug, io, net::SocketAddr};
+
+use super::io::internal::Io;
+use bytes::{Buf, Bytes};
+use tokio::prelude::*;
+use AddrInfo;
+
+/// A TcpStream where the initial reads will be served from `prefix`.
+#[derive(Debug)]
+pub struct Prefixed<S> {
+    prefix: Bytes,
+    io: S,
+}
+
+impl<S> Prefixed<S> {
+    pub fn new(prefix: Bytes, io: S) -> Self {
+        Self { prefix, io }
+    }
+}
+
+impl<S> io::Read for Prefixed<S> where S: Debug + io::Read {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, io::Error> {
+        // Check the length only once, since looking as the length
+        // of a Bytes isn't as cheap as the length of a &[u8].
+        let peeked_len = self.prefix.len();
+
+        if peeked_len == 0 {
+            self.io.read(buf)
+        } else {
+            let len = cmp::min(buf.len(), peeked_len);
+            buf[..len].copy_from_slice(&self.prefix.as_ref()[..len]);
+            self.prefix.advance(len);
+            // If we've finally emptied the peek_buf, drop it so we don't
+            // hold onto the allocated memory any longer. We won't peek
+            // again.
+            if peeked_len == len {
+                self.prefix = Default::default();
+            }
+            Ok(len)
+        }
+    }
+}
+
+impl<S> AsyncRead for Prefixed<S> where S: AsyncRead + Debug {
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        self.io.prepare_uninitialized_buffer(buf)
+    }
+}
+
+impl<S> io::Write for Prefixed<S> where S: Debug + io::Write {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.io.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.io.flush()
+    }
+}
+
+impl<S> AsyncWrite for Prefixed<S> where S: AsyncWrite + Debug {
+    fn shutdown(&mut self) -> Result<Async<()>, io::Error> {
+        self.io.shutdown()
+    }
```

Comment:
I *think* that because of `write_buf_erased`, it's fine to leave off the `write_buf` method, but it did trigger a tiny alarm in me when I noticed it missing...

Reply:
I'd prefer that we add it in case the erasure behavior changes in the future.

Reply:
Thanks. I added it.